File: diagonalOrder.java
//time complexity: O(n)
//space complexity: O(n)


class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        
        if(mat.length==0)
            return new int[0];
        int m=mat.length, n=mat[0].length, r=0,c=0;
        int[] result=new int[m*n];
        for(int i=0;i<result.length;i++)
        {
            result[i]=mat[r][c];
            if((r+c)%2==0)
            {
                if(c==n-1)
                {
                    r++;
                }
                else if(r==0)
                {
                    c++;
                }
                else
                {
                    r--;
                    c++;
                }
            }
            else
            {
                if(r==m-1)
                {
                    c++;
                }
                else if(c==0)
                {
                    r++;
                }
                else
                {
                    r++;
                    c--;
                }
            }
        }
          return result;
    
        
    }
}

File: Diagonal Order.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m =len(mat)
        n = len(mat[0])
        UP = 0
        DOWN = 1
        res = []
        row, col = 0,0
        resIndex = 0
        direction = UP
        
        while len(res) != m*n:
            res.append(mat[row][col])
            if direction == UP:
                
                if col == n-1:
                    row = row+1
                    direction = DOWN
                    
                    
                elif row == 0:
                    col = col+1
                    direction = DOWN
                    
                else:
                    row = row-1
                    col = col+1
            else:
                if row == m-1:
                    col = col+1
                    direction = UP
                    
                elif col == 0:
                    row = row+1
                    direction = UP
                else:
                    row = row+1
                    col = col-1
        return res
                
        